// @ideal-postcodes/address-finder/dist/contexts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toContextMap = exports.toContextList = exports.defaultContexts = void 0;
exports.defaultContexts = {
    GBR: {
        iso_2: "GB",
        iso_3: "GBR",
        emoji: "🇬🇧",
        rgeo: true,
        description: "United Kingdom",
    },
    USA: {
        iso_2: "US",
        iso_3: "USA",
        emoji: "🇺🇸",
        rgeo: false,
        description: "United States",
    },
};
const toContextList = (contexts, restrictCountries) => {
    const result = [];
    const codes = Object.keys(contexts);
    for (const code of codes) {
        if (restrictCountries.length > 0 &&
            !restrictCountries.some((e) => e === code))
            continue;
        result.push(contexts[code]);
    }
    result.sort((b, a) => b.description.localeCompare(a.description));
    return result;
};
exports.toContextList = toContextList;
const toContextMap = (contexts) => {
    const result = {};
    for (const context of contexts) {
        result[context.iso_3] = context;
    }
    return result;
};
exports.toContextMap = toContextMap;
