// @ideal-postcodes/address-finder/rollup.config.js
import commonjs from "@rollup/plugin-commonjs";
import resolve from "@rollup/plugin-node-resolve";
import ts from "rollup-plugin-ts";
import inject from "@rollup/plugin-inject";
import { dependencies } from "./package.json";

export const targets = ["last 2 versions"];

export const include = [
  ...Object.keys(dependencies).map((dep) => `node_modules/${dep}/**/*`),
  "node_modules/@ideal-postcodes/core-interface/**/*",
];

export const polyfills = {
  Promise: "promise-polyfill",
  Set: "core-js-pure/features/set",
  "Object.assign": "core-js-pure/features/object/assign",
};

export default [
  {
    input: "lib/index.ts",
    output: {
      file: "./address-finder.js",
      format: "umd",
      name: "IdealPostcodes",
      exports: "named",
    },
    plugins: [
      resolve({
        preferBuiltins: true,
        dedupe: ["@ideal-postcodes/jsutil"],
        mainFields: ["browser", "module", "main"],
        browser: true,
      }),
      commonjs(),
      inject(polyfills),
      ts({
        transpiler: "babel",
        browserslist: targets,
        include: ["lib/**/*", "node_modules/core-js-pure/**/*", ...include],
        babelConfig: {
          presets: [["@babel/preset-env", { targets }]],
          plugins: ["@babel/plugin-proposal-class-properties"],
        },
      }),
    ],
  },
];
